engines/ceg_prophecy_arrays.cpp

```diff
@@ -0,0 +1,459 @@
+** Copyright (c) 2019 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief An implementation of Counter-Example Guided Prophecy for array
+**        model checking. It is parameterized by an underlying model checking
+**        procedure which need not handle arrays (only UF). However, a common
+**        instantiation is with an IC3-style procedure, in which case we
+**        often refer to this algorithm as "prophic3".
+**
+**/
+
+#include <map>
+#include "assert.h"
+
+#include "engines/ceg_prophecy_arrays.h"
+#include "utils/logger.h"
+#include "utils/make_provers.h"
+
+using namespace smt;
+using namespace std;
+
+namespace pono {
+
+CegProphecyArrays::CegProphecyArrays(const Property & p,
+                                     Engine e,
+                                     smt::SolverEnum se)
+    : super(p, se),
+      conc_ts_(p.transition_system()),
+      solver_(conc_ts_.solver()),
+      abs_ts_(solver_),
+      e_(e),
+      abs_unroller_(abs_ts_, solver_),
+      aa_(conc_ts_, abs_ts_, true),
+      aae_(aa_, abs_unroller_, bad_, options_.cegp_axiom_red_),
+      pm_(abs_ts_)
+{
+  initialize();
+}
+
+CegProphecyArrays::CegProphecyArrays(const Property & p,
+                                     Engine e,
+                                     const SmtSolver & solver)
+    : super(p, solver),
+      conc_ts_(p.transition_system()),
+      solver_(conc_ts_.solver()),
+      abs_ts_(solver_),
+      e_(e),
+      abs_unroller_(abs_ts_, solver_),
+      aa_(conc_ts_, abs_ts_, true),
+      aae_(aa_, abs_unroller_, bad_, options_.cegp_axiom_red_),
+      pm_(abs_ts_)
+{
+  initialize();
+}
+
+CegProphecyArrays::CegProphecyArrays(const PonoOptions & opt,
+                                     const Property & p,
+                                     Engine e,
+                                     smt::SolverEnum se)
+    : super(opt, p, se),
+      conc_ts_(p.transition_system()),
+      solver_(conc_ts_.solver()),
+      abs_ts_(solver_),
+      e_(e),
+      abs_unroller_(abs_ts_, solver_),
+      aa_(conc_ts_, abs_ts_, true),
+      aae_(aa_, abs_unroller_, bad_, options_.cegp_axiom_red_),
+      pm_(abs_ts_)
+{
+  initialize();
+}
+
+CegProphecyArrays::CegProphecyArrays(const PonoOptions & opt,
+                                     const Property & p,
+                                     Engine e,
+                                     const smt::SmtSolver & solver)
+    : super(opt, p, solver),
+      conc_ts_(p.transition_system()),
+      solver_(conc_ts_.solver()),
+      abs_ts_(solver_),
+      abs_unroller_(abs_ts_, solver_),
+      e_(e),
+      aa_(conc_ts_, abs_ts_, true),
+      aae_(aa_, abs_unroller_, bad_, options_.cegp_axiom_red_),
+      pm_(abs_ts_)
+{
+  initialize();
+}
+
+ProverResult CegProphecyArrays::prove()
+{
```

Comment:
also, maybe use check_until with INT_MAX

Reply:
Makes sense! That means it will call `check_until(INT_MAX)` for the underlying model checker instead of `prove`. So far that means exactly the same thing, but I wasn't sure it would stay that way. I'll make the change, but just thought I'd point that out.